---
title: "TRTpred Demo 01"
author: "Rémy Pétremand"
date: '`r format(Sys.Date(), "%d %B, %Y")`'
output: 
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: false
    theme: cosmo
  fig_caption: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  tidy = FALSE, 
  cache = FALSE, 
  eval = TRUE,
  message = FALSE, 
  warning = FALSE)
```

# Getting Started

```{r}
library(TRTpred)  # TRTpred library (see installation instructions)
library(Seurat)   # Seurat library for the single-cell RNA and TCR sequencing objects
library(dplyr)    # dplyr library for data handeling
```

# Load demo data

From the TRTpred library we may load the scRNA-seq count matrix (UMI counts) and the corresponding meta.data containing the clone ids and their respective tumor-reactive labels. 

```{r}
# Load count matrix: "pt14.count.mat"
data("DemoCountMat", package = "TRTpred")
# Load metadata with scTCR-seq data and TRT labels:  pt14.metadata"
data("DemoMetadata", package = "TRTpred")

Seurat.Demo <- Seurat::CreateSeuratObject(counts = pt14.count.mat, meta.data = pt14.metadata)
```

In the metadata we may see that we have both the scTCR-seq data in the "CloneId" and the tumor-reactive labels in the "TumorReactiveGroundTruth" column. 

```{r}
reactable::reactable(Seurat.Demo@meta.data[1:5, c("orig.ident", "CloneId", "TumorReactiveGroundTruth")], rownames = F)
```

# Process data

To be able to use TRTpred we need to normalize the data and scale it. For the sake of simplicity here we only apply a standard log normalisation and scaling. 

```{r}
# Log Normalize
Seurat.Demo <- NormalizeData(Seurat.Demo, normalization.method = "LogNormalize", scale.factor = 1E4, verbose = F)

# Scaling
Seurat.Demo <- ScaleData(Seurat.Demo, features = rownames(Seurat.Demo), verbose = F)
```

# Apply TRTpred

We can obtain the prediction from TRTpred directly by using the `GetTRTpred()` function. 

```{r}
# Get the cell-wise TRT prediction
TRTpred.prediction <- GetTRTpred(seurat.obj = Seurat.Demo, seurat.assay = "RNA", seurat.slot = "scale.data")

# Map the clone-id columns to the cell-wise predictions data.frame
TRTpred.prediction$CloneId <- Seurat.Demo@meta.data[rownames(TRTpred.prediction), ]$CloneId

# Map the ground-truth column to the cell-wise predictions data.frame
TRTpred.prediction$TumorReactiveGroundTruth <- Seurat.Demo@meta.data[rownames(TRTpred.prediction), ]$TumorReactiveGroundTruth

# show results: 
reactable::reactable(TRTpred.prediction[1:5, ])
```

The results are the scaled scores (`score` column) and Boolean prediction (`pred` columns, TRUE is tumor reactive) per cells. To obtain clone-wise prediction 

```{r}
# Get the clone-wise TRT prediction
TRTpred.prediction.clone <- GetClonePred(pred.df = TRTpred.prediction)

# show results: 
reactable::reactable(TRTpred.prediction.clone[1:5, ])
```

# Evaluation of the prediction

You may obtain the evaluation metrics (accuracy, AUC, sensitivity, ...) using the function `GetAccuracyMetrics()`

```{r}
evaluation.res <- TRTpred::GetAccuracyMetrics(
  ground_truth = TRTpred.prediction$TumorReactiveGroundTruth, 
  preds = TRTpred.prediction$pred, 
  scores = TRTpred.prediction$score,
  metrics = c("auc", "accuracy", "sensitivity"))
```

The metrics are: 

```{r}
evaluation.res
```



